src/swapRouter.ts

```diff
@@ -175,62 +180,203 @@ export abstract class SwapRouter {
     return calldatas
   }
 
+  /**
+   * @notice Generates the calldata for a MixedRouteSwap. Since MixedRoutes cannot be singleHops, we will instead generate
+   *         them via the existing encodeV3Swap and encodeV2Swap methods.
+   * @param trade The MixedRouteTrade to encode.
+   * @param options
+   * @param routerMustCustody
+   * @param performAggregatedSlippageCheck
+   * @returns
+   */
+  private static encodeMixedRouteSwap(
+    trade: MixedRouteTrade<Currency, Currency, TradeType>,
+    options: SwapOptions,
+    routerMustCustody: boolean,
+    performAggregatedSlippageCheck: boolean
+  ): string[] {
+    const calldatas: string[] = []
+
+    invariant(trade.tradeType === TradeType.EXACT_INPUT, 'TRADE_TYPE')
+
+    for (const { route, inputAmount, outputAmount } of trade.swaps) {
+      const amountIn: string = toHex(trade.maximumAmountIn(options.slippageTolerance, inputAmount).quotient)
+      const amountOut: string = toHex(trade.minimumAmountOut(options.slippageTolerance, outputAmount).quotient)
+
+      // flag for whether the trade is single hop or not
+      const singleHop = route.pools.length === 1
+
+      const recipient = routerMustCustody
+        ? ADDRESS_THIS
+        : typeof options.recipient === 'undefined'
+        ? MSG_SENDER
+        : validateAndParseAddress(options.recipient)
+
+      const mixedRouteIsAllV3 = (route: MixedRoute<Currency, Currency>) => {
+        return route.pools.every((pool) => pool instanceof Pool)
+      }
+
+      if (singleHop) {
+        /// For single hop, since it isn't really a mixedRoute, we'll just mimic behavior of V3 or V2
+        /// We don't use encodeV3Swap or encodeV2Swap because casting the trade to a V3Trade or V2Trade is overcomplicated
+        if (route.pools.every((pool) => pool instanceof Pool)) {
+          const exactInputSingleParams = {
+            tokenIn: route.path[0].address,
+            tokenOut: route.path[1].address,
+            fee: (route.pools as Pool[])[0].fee,
+            recipient,
+            amountIn,
+            amountOutMinimum: performAggregatedSlippageCheck ? 0 : amountOut,
+            sqrtPriceLimitX96: 0,
+          }
+
+          calldatas.push(SwapRouter.INTERFACE.encodeFunctionData('exactInputSingle', [exactInputSingleParams]))
+        } else {
+          const path = trade.route.path.map((token) => token.address)
+
+          const exactInputParams = [amountIn, performAggregatedSlippageCheck ? 0 : amountOut, path, recipient]
+
+          calldatas.push(SwapRouter.INTERFACE.encodeFunctionData('swapExactTokensForTokens', exactInputParams))
+        }
+      } else {
+        const sections = divideMixedRouteIntoConsecutiveSections(route)
+
+        const firstInputToken = route.input.wrapped
+        let outputToken = firstInputToken
```

Comment:
this is confusing. could just assign inputToken to route.input.wrapped if it's the first token in the section. and then you can do the updates to inputToken, outputToken after the call to `new MixedRouteSDK`

Reply:
oh wow yeah I agree that's confusing. I think the logic was more complex before but now it doesn't rly make sense.

Reply:
Now setting `inputToken` to route.input.wrapped, and assigning `outputToken` before call to `new MixedRouteSDK` and then setting `inputToken = outputToken` after route creation